src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -257,6 +258,338 @@ private struct IntPtrAndHandle
+            }
+        }
+
+        public static unsafe string? MakeMarshalSignatureInfo (IntPtr typePtr, IntPtr methodPtr) {
+            var mb = MethodFromPointers(typePtr, methodPtr);
+            if (mb == null)
+                return null;
+
+            var result = "{ " +
+                $"\"result\": {MakeMarshalTypeRecord((mb as MethodInfo)?.ReturnType ?? typeof(void))}, " +
+                $"\"typePtr\": {typePtr}, \"methodPtr\": {methodPtr}, " +
+                "\"parameters\": [";
+
+            int i = 0;
+            foreach (var p in mb.GetParameters()) {
+                if (i > 0)
+                    result += ", ";
+                result += MakeMarshalTypeRecord(p.ParameterType);
+                i++;
+            }
+
+            result += "] }";
+
+            return result;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "Trimming doesn't affect types eligible for marshalling. Different exception for invalid inputs doesn't matter.")]
+        private static unsafe string GetAndEscapeStringProperty (Type type, string name) {
+            var info = type.GetProperty(
+                name, BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic
+            );
+
+            var value = info?.GetValue(null) as string;
+            if (value == null)
+                return "null";
+
+            var result = "\"";
+            foreach (var ch in value) {
+                switch (ch) {
+                    case '\'':
+                        result += '\'';
+                        continue;
+                    case '"':
+                        result += '\"';
+                        continue;
+                    case '\\':
+                        result += "\\\\";
+                        continue;
+                    case '\n':
+                        result += "\\n";
+                        continue;
+                }
+
+                if (ch < ' ') {
+                    result += "\\x";
+                    if (ch <= 0xF)
+                        result += '0';
+                    result += ((int)ch).ToString("X");
+                } else {
+                    result += ch;
+                }
+            }
+            result += '"';
+
+            return result;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "Trimming doesn't affect types eligible for marshalling. Different exception for invalid inputs doesn't matter.")]
+        private static unsafe IntPtr GetMarshalMethodPointer (Type type, string name, out Type? returnType, out Type? parameterType) {
+            var info = type.GetMethod(
+                name, BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic
+            );
+            if (info == null) {
+                parameterType = null;
```

Comment:
Agreed, though note that everything JS touches needs to have nullable arguments, because JS can hand us null ptrs no matter what our signature is. (We could implement some sort of check to enforce this if we figure out a way for JS to know that parameters aren't nullable?)